driver/src/driver/scheduler/system.rs

```diff
@@ -62,92 +65,131 @@ impl<'a> SystemScheduler<'a> {
         }
     }
 
-    /// Returns how long to sleep before starting the next iteration.
-    fn run_once(&mut self, now: SystemTime) -> Duration {
-        const RETRY_TIMEOUT: Duration = Duration::from_secs(10);
-        match self.determine_action(now) {
-            Ok(Action::Sleep(duration)) => {
-                info!("Sleeping {}s.", duration.as_secs());
-                duration
-            }
-            Ok(Action::Solve(batch_id, time_limit)) => {
-                info!(
-                    "Starting to solve batch {} with time limit {}s.",
-                    batch_id.0,
-                    time_limit.as_secs_f64()
-                );
-                match self.driver.run(batch_id.0.into(), time_limit) {
-                    DriverResult::Ok => {
-                        info!("Batch {} solved successfully.", batch_id.0);
-                        self.last_solved_batch.replace(batch_id);
-                        Duration::from_secs(0)
-                    }
-                    DriverResult::Retry(err) => {
-                        error!("StableXDriver retryable error: {}", err);
-                        RETRY_TIMEOUT
-                    }
-                    DriverResult::Skip(err) => {
-                        error!("StableXDriver unretryable error: {}", err);
-                        self.last_solved_batch.replace(batch_id);
-                        Duration::from_secs(0)
-                    }
-                }
-            }
-            Err(err) => {
-                error!("Scheduler error: {}", err);
-                RETRY_TIMEOUT
+    fn start_solving_in_thread<'b>(
+        &self,
+        batch_id: BatchId,
+        solver_time_limit: Duration,
+        scope: &Scope<'b>,
+    ) where
+        'a: 'b,
+    {
+        let driver = self.driver;
+        let auction_timing_configuration = self.auction_timing_configuration;
+        scope.spawn(move |_| loop {
```

Comment:
It's not a do while loop, just kind of. For example you cannot `break` and `continue`.

Reply:
If `should_attempt_to_solve_again` is correct then the loop is bounded. I don't find a counter helpful.

Reply:
I just feel like we are adding a lot of potentially unbounded loops into the codebase, which makes it harder to reason about how and when things will terminate (we have quite some nested branches here and in `should_attempt_to_solve_again` that need to be understood).

On that note, would it make sense to add some metrics around how many threads we have and how long they run on average?

Reply:
We still expect there to only be one thread unless there is a bug with timeouts or the solver. A number of running threads metric would help up us notice if such a bug occurs because we would observe more than 1 thread.

Thread runtime is not useful because it is already measured by the processing times (and it would have the same problems as them if there were multiple threads running at the same time).